Dependabot auto merge failing

The dependabot automerge workflow is still failing when referenced from another repo.

Error log:

~~~
Run set -euo pipefail
  set -euo pipefail
  if [[ -n "91" ]]; then
    export INPUT_PULL_REQUEST_NUMBER="91"
  fi
  if [[ -n "" ]]; then
    export INPUT_REPOSITORY=""
  fi
  if [[ "${ACT:-}" == "true" ]]; then
    export UV_PROJECT_ENVIRONMENT="/tmp/uv-project"
    export UV_CACHE_DIR="/tmp/uv-cache"
  fi
  uv run --script "/home/runner/work/pg-embedded-setup-unpriv/pg-embedded-setup-unpriv/workflow-src/workflow_scripts/dependabot_automerge.py"
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    UV_PYTHON: 3.13
    UV_CACHE_DIR: /home/runner/work/_temp/setup-uv-cache
    INPUT_GITHUB_TOKEN: ***
    INPUT_REQUIRED_LABEL: dependencies
    INPUT_MERGE_METHOD: squash
    INPUT_DRY_RUN: false
Downloading cpython-3.13.11-linux-x86_64-gnu (download) (33.4MiB)
 Downloaded cpython-3.13.11-linux-x86_64-gnu (download)
/home/runner/.local/share/uv/python/cpython-3.13.11-linux-x86_64-gnu/bin/python3.13: can't open file '/home/runner/work/pg-embedded-setup-unpriv/pg-embedded-setup-unpriv/workflow-src/workflow_scripts/dependabot_automerge.py': [Errno 2] No such file or directory
Error: Process completed with exit code 2.
~~~

Referencing workflow (in `leynos/pg-embedded-setup-unpriv`):

~~~
name: dependabot-automerge

# Uses pull_request_target to enable auto-merge with write permissions.
# Safe because the reusable workflow never checks out or executes PR code;
# it only reads event metadata and makes GitHub API calls.

on:
  pull_request_target:
    branches: [main]
    types: [opened, reopened, synchronize, labeled, ready_for_review]
  workflow_dispatch:
    inputs:
      pull-request-number:
        type: number
        required: true

jobs:
  automerge:
    permissions:
      contents: write
      pull-requests: write
      checks: read
      statuses: read
    if: ${{ github.event_name == 'workflow_dispatch' || github.actor == 'dependabot[bot]' }}
    uses: leynos/shared-actions/.github/workflows/dependabot-automerge.yml@98246364002def859ab550fe01ab3bca55dcfd61
    with:
      pull-request-number: ${{ inputs.pull-request-number || github.event.pull_request.number }}
~~~

--- a/.github/actions/rust-build-release/tests/conftest.py
+++ b/.github/actions/rust-build-release/tests/conftest.py
@@ -333,12 +333,22 @@ def ensure_toolchain_ready() -> cabc.Callable[[str, str], None]:
             pytest.skip("rustup not installed")
         if not Path(rustup_path).is_file():
             pytest.skip(f"rustup missing at resolved path: {rustup_path}")
+        rustup_bin = Path(rustup_path).resolve()
+        cargo_home = rustup_bin.parents[1]
+        if not cargo_home.is_dir():
+            pytest.skip(f"cargo home missing for rustup: {cargo_home}")
+        rustup_home = cargo_home.parent / ".rustup"
+        rustup_env = {
+            "CARGO_HOME": str(cargo_home),
+            "RUSTUP_HOME": str(rustup_home),
+        }
         try:
             _, stdout, _ = typ.cast(
                 "tuple[int, str, str]",
                 run_cmd(
                     local[rustup_path]["toolchain", "list"],
                     method="run",
+                    env=rustup_env,
                 ),
             )
         except (OSError, ProcessExecutionError) as exc:  # pragma: no cover - env guard
@@ -363,7 +373,8 @@ def ensure_toolchain_ready() -> cabc.Callable[[str, str], None]:
                     install_spec,
                     "--profile",
                     "minimal",
-                ]
+                ],
+                env=rustup_env,
             )
 
     return _ensure

--- a/.github/actions/rust-build-release/tests/conftest.py
+++ b/.github/actions/rust-build-release/tests/conftest.py
@@ -8,6 +8,7 @@ import importlib.util
 import shutil
 import sys
 import typing as typ
+from dataclasses import dataclass  # noqa: ICN003
 from pathlib import Path
 
 import pytest
@@ -367,24 +368,32 @@ def _get_installed_toolchains(
     return [line.split()[0] for line in stdout.splitlines() if line.strip()]
 
 
+@dataclass(frozen=True)
+class ToolchainSpec:
+    """Specification for a Rust toolchain installation."""
+
+    version: str
+    host_target: str
+
+
 def _install_toolchain_if_needed(
     *,
     rustup_path: str,
     rustup_env: dict[str, str],
-    toolchain_version: str,
-    host_target: str,
+    toolchain_spec: ToolchainSpec,
     installed_names: cabc.Sequence[str],
 ) -> None:
     """Install the requested toolchain when it is missing."""
     expected = {
-        toolchain_version,
-        f"{toolchain_version}-{host_target}",
+        toolchain_spec.version,
+        f"{toolchain_spec.version}-{toolchain_spec.host_target}",
     }
     if all(name not in expected for name in installed_names):
         install_spec = (
-            f"{toolchain_version}-{host_target}"
-            if sys.platform == "win32" and host_target.endswith("-pc-windows-gnu")
-            else toolchain_version
+            f"{toolchain_spec.version}-{toolchain_spec.host_target}"
+            if sys.platform == "win32"
+            and toolchain_spec.host_target.endswith("-pc-windows-gnu")
+            else toolchain_spec.version
         )
         run_cmd(
             local[rustup_path][
@@ -410,8 +419,10 @@ def ensure_toolchain_ready() -> cabc.Callable[[str, str], None]:
         _install_toolchain_if_needed(
             rustup_path=rustup_bin.as_posix(),
             rustup_env=rustup_env,
-            toolchain_version=toolchain_version,
-            host_target=host_target,
+            toolchain_spec=ToolchainSpec(
+                version=toolchain_version,
+                host_target=host_target,
+            ),
             installed_names=installed_names,
         )
 

--- a/.github/actions/rust-build-release/tests/conftest.py
+++ b/.github/actions/rust-build-release/tests/conftest.py
@@ -3,12 +3,12 @@
 from __future__ import annotations
 
 import collections.abc as cabc
+import dataclasses
 import importlib
 import importlib.util
 import shutil
 import sys
 import typing as typ
-from dataclasses import dataclass  # noqa: ICN003
 from pathlib import Path
 
 import pytest
@@ -368,7 +368,7 @@ def _get_installed_toolchains(
     return [line.split()[0] for line in stdout.splitlines() if line.strip()]
 
 
-@dataclass(frozen=True)
+@dataclasses.dataclass(frozen=True)
 class ToolchainSpec:
     """Specification for a Rust toolchain installation."""
 

--- a/.github/actions/rust-build-release/tests/conftest.py
+++ b/.github/actions/rust-build-release/tests/conftest.py
@@ -334,10 +334,14 @@ def ensure_toolchain_ready() -> cabc.Callable[[str, str], None]:
         if not Path(rustup_path).is_file():
             pytest.skip(f"rustup missing at resolved path: {rustup_path}")
         rustup_bin = Path(rustup_path).resolve()
+        if len(rustup_bin.parents) < 2:
+            pytest.skip(f"rustup path lacks expected parents: {rustup_bin}")
         cargo_home = rustup_bin.parents[1]
         if not cargo_home.is_dir():
             pytest.skip(f"cargo home missing for rustup: {cargo_home}")
         rustup_home = cargo_home.parent / ".rustup"
+        if not rustup_home.is_dir():
+            pytest.skip(f"rustup home missing for rustup: {rustup_home}")
         rustup_env = {
             "CARGO_HOME": str(cargo_home),
             "RUSTUP_HOME": str(rustup_home),

--- a/.github/actions/rust-build-release/tests/conftest.py
+++ b/.github/actions/rust-build-release/tests/conftest.py
@@ -335,19 +335,18 @@ def _validate_rustup_binary(rustup_path: str | None) -> Path:
     return rustup_bin
 
 
-def _setup_rust_environment(rustup_bin: Path) -> tuple[Path, Path, dict[str, str]]:
-    """Return cargo/rustup paths and a runtime environment mapping."""
+def _setup_rust_environment(rustup_bin: Path) -> dict[str, str]:
+    """Return a runtime environment mapping for rustup."""
     cargo_home = rustup_bin.parents[1]
     if not cargo_home.is_dir():
         pytest.skip(f"cargo home missing for rustup: {cargo_home}")
     rustup_home = cargo_home.parent / ".rustup"
     if not rustup_home.is_dir():
         pytest.skip(f"rustup home missing for rustup: {rustup_home}")
-    rustup_env = {
+    return {
         "CARGO_HOME": str(cargo_home),
         "RUSTUP_HOME": str(rustup_home),
     }
-    return cargo_home, rustup_home, rustup_env
 
 
 def _get_installed_toolchains(
@@ -406,7 +405,7 @@ def ensure_toolchain_ready() -> cabc.Callable[[str, str], None]:
     def _ensure(toolchain_version: str, host_target: str) -> None:
         rustup_path = shutil.which("rustup")
         rustup_bin = _validate_rustup_binary(rustup_path)
-        _, _, rustup_env = _setup_rust_environment(rustup_bin)
+        rustup_env = _setup_rust_environment(rustup_bin)
         installed_names = _get_installed_toolchains(rustup_bin.as_posix(), rustup_env)
         _install_toolchain_if_needed(
             rustup_path=rustup_bin.as_posix(),

--- a/.github/actions/rust-build-release/tests/conftest.py
+++ b/.github/actions/rust-build-release/tests/conftest.py
@@ -323,63 +323,98 @@ def patch_common_main_deps(
     return harness
 
 
+def _validate_rustup_binary(rustup_path: str | None) -> Path:
+    """Validate and return the resolved rustup binary path."""
+    if rustup_path is None:  # pragma: no cover - guarded by caller checks
+        pytest.skip("rustup not installed")
+    if not Path(rustup_path).is_file():
+        pytest.skip(f"rustup missing at resolved path: {rustup_path}")
+    rustup_bin = Path(rustup_path).resolve()
+    if len(rustup_bin.parents) < 2:
+        pytest.skip(f"rustup path lacks expected parents: {rustup_bin}")
+    return rustup_bin
+
+
+def _setup_rust_environment(rustup_bin: Path) -> tuple[Path, Path, dict[str, str]]:
+    """Return cargo/rustup paths and a runtime environment mapping."""
+    cargo_home = rustup_bin.parents[1]
+    if not cargo_home.is_dir():
+        pytest.skip(f"cargo home missing for rustup: {cargo_home}")
+    rustup_home = cargo_home.parent / ".rustup"
+    if not rustup_home.is_dir():
+        pytest.skip(f"rustup home missing for rustup: {rustup_home}")
+    rustup_env = {
+        "CARGO_HOME": str(cargo_home),
+        "RUSTUP_HOME": str(rustup_home),
+    }
+    return cargo_home, rustup_home, rustup_env
+
+
+def _get_installed_toolchains(
+    rustup_path: str, rustup_env: dict[str, str]
+) -> list[str]:
+    """Return installed toolchain names from rustup."""
+    try:
+        _, stdout, _ = typ.cast(
+            "tuple[int, str, str]",
+            run_cmd(
+                local[rustup_path]["toolchain", "list"],
+                method="run",
+                env=rustup_env,
+            ),
+        )
+    except (OSError, ProcessExecutionError) as exc:  # pragma: no cover - env guard
+        pytest.skip(f"rustup unavailable on runner: {exc}")
+    return [line.split()[0] for line in stdout.splitlines() if line.strip()]
+
+
+def _install_toolchain_if_needed(
+    *,
+    rustup_path: str,
+    rustup_env: dict[str, str],
+    toolchain_version: str,
+    host_target: str,
+    installed_names: cabc.Sequence[str],
+) -> None:
+    """Install the requested toolchain when it is missing."""
+    expected = {
+        toolchain_version,
+        f"{toolchain_version}-{host_target}",
+    }
+    if all(name not in expected for name in installed_names):
+        install_spec = (
+            f"{toolchain_version}-{host_target}"
+            if sys.platform == "win32" and host_target.endswith("-pc-windows-gnu")
+            else toolchain_version
+        )
+        run_cmd(
+            local[rustup_path][
+                "toolchain",
+                "install",
+                install_spec,
+                "--profile",
+                "minimal",
+            ],
+            env=rustup_env,
+        )
+
+
 @pytest.fixture
 def ensure_toolchain_ready() -> cabc.Callable[[str, str], None]:
     """Return a helper that ensures the requested toolchain is installed."""
 
     def _ensure(toolchain_version: str, host_target: str) -> None:
         rustup_path = shutil.which("rustup")
-        if rustup_path is None:  # pragma: no cover - guarded by caller checks
-            pytest.skip("rustup not installed")
-        if not Path(rustup_path).is_file():
-            pytest.skip(f"rustup missing at resolved path: {rustup_path}")
-        rustup_bin = Path(rustup_path).resolve()
-        if len(rustup_bin.parents) < 2:
-            pytest.skip(f"rustup path lacks expected parents: {rustup_bin}")
-        cargo_home = rustup_bin.parents[1]
-        if not cargo_home.is_dir():
-            pytest.skip(f"cargo home missing for rustup: {cargo_home}")
-        rustup_home = cargo_home.parent / ".rustup"
-        if not rustup_home.is_dir():
-            pytest.skip(f"rustup home missing for rustup: {rustup_home}")
-        rustup_env = {
-            "CARGO_HOME": str(cargo_home),
-            "RUSTUP_HOME": str(rustup_home),
-        }
-        try:
-            _, stdout, _ = typ.cast(
-                "tuple[int, str, str]",
-                run_cmd(
-                    local[rustup_path]["toolchain", "list"],
-                    method="run",
-                    env=rustup_env,
-                ),
-            )
-        except (OSError, ProcessExecutionError) as exc:  # pragma: no cover - env guard
-            pytest.skip(f"rustup unavailable on runner: {exc}")
-        installed_names = [
-            line.split()[0] for line in stdout.splitlines() if line.strip()
-        ]
-        expected = {
-            toolchain_version,
-            f"{toolchain_version}-{host_target}",
-        }
-        if all(name not in expected for name in installed_names):
-            install_spec = (
-                f"{toolchain_version}-{host_target}"
-                if sys.platform == "win32" and host_target.endswith("-pc-windows-gnu")
-                else toolchain_version
-            )
-            run_cmd(
-                local[rustup_path][
-                    "toolchain",
-                    "install",
-                    install_spec,
-                    "--profile",
-                    "minimal",
-                ],
-                env=rustup_env,
-            )
+        rustup_bin = _validate_rustup_binary(rustup_path)
+        _, _, rustup_env = _setup_rust_environment(rustup_bin)
+        installed_names = _get_installed_toolchains(rustup_bin.as_posix(), rustup_env)
+        _install_toolchain_if_needed(
+            rustup_path=rustup_bin.as_posix(),
+            rustup_env=rustup_env,
+            toolchain_version=toolchain_version,
+            host_target=host_target,
+            installed_names=installed_names,
+        )
 
     return _ensure
 

--- a/.github/workflows/dependabot-automerge.yml
+++ b/.github/workflows/dependabot-automerge.yml
@@ -36,6 +36,7 @@ jobs:
         shell: bash
         env:
           WORKFLOW_REF: ${{ github.workflow_ref }}
+          WORKFLOW_SHA: ${{ github.workflow_sha }}
           REPO: ${{ github.repository }}
           REF: ${{ github.ref }}
         run: |
@@ -48,7 +49,10 @@ jobs:
             echo "ref=" >> "${GITHUB_OUTPUT}"
             exit 0
           fi
-          if [[ -n "${WORKFLOW_REF}" ]]; then
+          if [[ -n "${WORKFLOW_SHA}" ]]; then
+            repo="${WORKFLOW_REF%%/.github/workflows/*}"
+            ref="${WORKFLOW_SHA}"
+          elif [[ -n "${WORKFLOW_REF}" ]]; then
             repo="${WORKFLOW_REF%%/.github/workflows/*}"
             ref="${WORKFLOW_REF##*@}"
           else
@@ -59,6 +63,13 @@ jobs:
           echo "workflow_dir=${workspace}/workflow-src" >> "${GITHUB_OUTPUT}"
           echo "repo=${repo}" >> "${GITHUB_OUTPUT}"
           echo "ref=${ref}" >> "${GITHUB_OUTPUT}"
+      - name: Debug workflow source
+        shell: bash
+        run: |
+          set -euo pipefail
+          echo "workflow repo: ${{ steps.workflow-source.outputs.repo }}"
+          echo "workflow ref: ${{ steps.workflow-source.outputs.ref }}"
+          echo "workflow dir: ${{ steps.workflow-source.outputs.workflow_dir }}"
       - name: Checkout workflow repository
         if: ${{ steps.workflow-source.outputs.checkout == 'true' }}
         # v4
@@ -67,7 +78,19 @@ jobs:
           repository: ${{ steps.workflow-source.outputs.repo }}
           ref: ${{ steps.workflow-source.outputs.ref }}
           path: workflow-src
-          fetch-depth: 1
+          fetch-depth: 0
+      - name: Verify workflow checkout
+        if: ${{ steps.workflow-source.outputs.checkout == 'true' }}
+        shell: bash
+        run: |
+          set -euo pipefail
+          script_path="workflow-src/workflow_scripts/dependabot_automerge.py"
+          if [[ ! -f "${script_path}" ]]; then
+            echo "Missing ${script_path} after checkout."
+            echo "workflow-src contents:"
+            ls -la workflow-src
+            exit 1
+          fi
       - name: Install uv (act)
         if: ${{ env.ACT == 'true' }}
         shell: bash

--- a/.github/workflows/dependabot-automerge.yml
+++ b/.github/workflows/dependabot-automerge.yml
@@ -50,7 +50,13 @@ jobs:
             exit 0
           fi
           if [[ -n "${WORKFLOW_SHA}" ]]; then
-            repo="${WORKFLOW_REF%%/.github/workflows/*}"
+            repo="${REPO}"
+            if [[ -n "${WORKFLOW_REF}" && "${WORKFLOW_REF}" == */.github/workflows/*@* ]]; then
+              parsed_repo="${WORKFLOW_REF%%/.github/workflows/*}"
+              if [[ -n "${parsed_repo}" ]]; then
+                repo="${parsed_repo}"
+              fi
+            fi
             ref="${WORKFLOW_SHA}"
           elif [[ -n "${WORKFLOW_REF}" ]]; then
             repo="${WORKFLOW_REF%%/.github/workflows/*}"
@@ -78,7 +84,7 @@ jobs:
           repository: ${{ steps.workflow-source.outputs.repo }}
           ref: ${{ steps.workflow-source.outputs.ref }}
           path: workflow-src
-          fetch-depth: 0
+          fetch-depth: 1
       - name: Verify workflow checkout
         if: ${{ steps.workflow-source.outputs.checkout == 'true' }}
         shell: bash

--- a/.github/workflows/dependabot-automerge.yml
+++ b/.github/workflows/dependabot-automerge.yml
@@ -84,11 +84,14 @@ jobs:
         shell: bash
         run: |
           set -euo pipefail
-          script_path="workflow-src/workflow_scripts/dependabot_automerge.py"
+          workflow_dir="${{ steps.workflow-source.outputs.workflow_dir }}"
+          script_path="${workflow_dir}/workflow_scripts/dependabot_automerge.py"
           if [[ ! -f "${script_path}" ]]; then
             echo "Missing ${script_path} after checkout."
-            echo "workflow-src contents:"
-            ls -la workflow-src
+            echo "Checked-out commit:"
+            git -C "${workflow_dir}" rev-parse HEAD
+            echo "Workflow directory contents:"
+            ls -la "${workflow_dir}"
             exit 1
           fi
       - name: Install uv (act)

--- a/.github/workflows/dependabot-automerge.yml
+++ b/.github/workflows/dependabot-automerge.yml
@@ -98,6 +98,11 @@ jobs:
             git -C "${workflow_dir}" rev-parse HEAD
             echo "Workflow directory contents:"
             ls -la "${workflow_dir}"
+            workflow_scripts_dir="${workflow_dir}/workflow_scripts"
+            if [[ -d "${workflow_scripts_dir}" ]]; then
+              echo "Workflow scripts contents:"
+              ls -la "${workflow_scripts_dir}"
+            fi
             exit 1
           fi
       - name: Install uv (act)

